indexing-service/src/main/java/io/druid/indexing/overlord/RemoteTaskRunner.java

```diff
@@ -880,55 +888,79 @@ private void removeWorker(final Worker worker)
   {
-    removedWorkerCleanups.put(
-        worker, cleanupExec.schedule(
-            new Runnable()
-            {
-              @Override
-              public void run()
-              {
-                log.info("Running scheduled cleanup for Worker[%s]", worker);
-                try {
-                  for (String assignedTask : tasksToFail) {
-                    String taskPath = JOINER.join(indexerZkConfig.getTasksPath(), worker, assignedTask);
-                    String statusPath = JOINER.join(indexerZkConfig.getStatusPath(), worker, assignedTask);
-                    if (cf.checkExists().forPath(taskPath) != null) {
-                      cf.delete().guaranteed().forPath(taskPath);
-                    }
+    // This method is only called from the PathChildrenCache event handler, so this may look like a race,
+    // but is actually not.
+    cancelWorkerCleanup(worker);
 
-                    if (cf.checkExists().forPath(statusPath) != null) {
-                      cf.delete().guaranteed().forPath(statusPath);
-                    }
-
-                    log.info("Failing task[%s]", assignedTask);
-                    RemoteTaskRunnerWorkItem taskRunnerWorkItem = runningTasks.remove(assignedTask);
-                    if (taskRunnerWorkItem != null) {
-                      taskRunnerWorkItem.setResult(TaskStatus.failure(taskRunnerWorkItem.getTaskId()));
-                    } else {
-                      log.warn("RemoteTaskRunner has no knowledge of task[%s]", assignedTask);
-                    }
-                  }
-
-                  // worker is gone, remove worker task status announcements path.
-                  String workerStatusPath = JOINER.join(indexerZkConfig.getStatusPath(), worker);
-                  if (cf.checkExists().forPath(workerStatusPath) != null) {
-                    cf.delete().guaranteed().forPath(JOINER.join(indexerZkConfig.getStatusPath(), worker));
-                  }
+    final ListenableScheduledFuture<?> cleanupTask = cleanupExec.schedule(
+        new Runnable()
+        {
+          @Override
+          public void run()
+          {
+            log.info("Running scheduled cleanup for Worker[%s]", worker);
+            try {
+              for (String assignedTask : tasksToFail) {
+                String taskPath = JOINER.join(indexerZkConfig.getTasksPath(), worker, assignedTask);
+                String statusPath = JOINER.join(indexerZkConfig.getStatusPath(), worker, assignedTask);
+                if (cf.checkExists().forPath(taskPath) != null) {
+                  cf.delete().guaranteed().forPath(taskPath);
                 }
-                catch (Exception e) {
-                  log.makeAlert("Exception while cleaning up worker[%s]", worker).emit();
-                  throw Throwables.propagate(e);
+
+                if (cf.checkExists().forPath(statusPath) != null) {
+                  cf.delete().guaranteed().forPath(statusPath);
```

Comment:
There shouldn't be one in practice, because nothing should be removing these other than the RTR and it has a leader election thing going on.

I'll change it though to check for that, and also drop the `.guaranteed()` since it's not useful and may actually just cause problems.